packages/cli/src/execution-lifecycle/execution-lifecycle-hooks.ts

```diff
@@ -34,6 +34,61 @@ import {
 } from './shared/shared-hook-functions';
 import { toSaveSettings } from './to-save-settings';
 
+function mergeHookFunctions(...hookFunctions: IWorkflowExecuteHooks[]): IWorkflowExecuteHooks {
+	const result: IWorkflowExecuteHooks = {
+		nodeExecuteBefore: [],
+		nodeExecuteAfter: [],
+		workflowExecuteBefore: [],
+		workflowExecuteAfter: [],
+		sendResponse: [],
+		nodeFetchedData: [],
+	};
+	for (const hooks of hookFunctions) {
+		for (const key in hooks) {
+			if (hooks.hasOwnProperty(key)) {
+				result[key]!.push(...hooks[key]!);
+			}
+		}
+	}
+	return result;
+}
+
+function hookFunctionsWorkflowTelemetry(userId?: string): IWorkflowExecuteHooks {
+	const eventService = Container.get(EventService);
+	return {
+		workflowExecuteBefore: [
+			async function (this: WorkflowHooks): Promise<void> {
+				const { executionId, workflowData } = this;
+				eventService.emit('workflow-pre-execute', { executionId, data: workflowData });
+			},
+		],
+		workflowExecuteAfter: [
+			async function (this: WorkflowHooks, runData: IRun): Promise<void> {
+				const { executionId, workflowData: workflow } = this;
+				eventService.emit('workflow-post-execute', { executionId, runData, workflow, userId });
+			},
+		],
+	};
+}
+
+function hookFunctionsNodeTelemetry(): IWorkflowExecuteHooks {
+	const eventService = Container.get(EventService);
+	return {
+		nodeExecuteBefore: [
+			async function (this: WorkflowHooks, nodeName: string): Promise<void> {
+				const { executionId, workflowData: workflow } = this;
+				eventService.emit('node-pre-execute', { executionId, workflow, nodeName });
+			},
+		],
+		nodeExecuteAfter: [
+			async function (this: WorkflowHooks, nodeName: string): Promise<void> {
+				const { executionId, workflowData: workflow } = this;
+				eventService.emit('node-post-execute', { executionId, workflow, nodeName });
+			},
+		],
```

Comment:
Do these need to be async? Same elsewhere.

Reply:
the async is coming from the typing. I'll change this in the WorkflowHooks refactor.